src/treewidth.jl

```diff
@@ -0,0 +1,166 @@
+    eo = copy(elimination_order.order)
+    incidence_list = copy(incidence_list)
+    contraction_tree_nodes = Vector{Union{VT, ContractionTree}}(collect(keys(incidence_list.v2e)))
+    tensors_list = Dict{VT, Int}()
+    for (i, v) in enumerate(contraction_tree_nodes)
+        tensors_list[v] = i
+    end
+
+    flag = contraction_tree_nodes[1]
+
+    while !isempty(eo)
+        e = pop!(eo)
+        if haskey(incidence_list.e2v, e)
+            vs = incidence_list.e2v[e]
+            if length(vs) >= 2
+                sub_list = IncidenceList(Dict([v => incidence_list.v2e[v] for v in vs]); openedges=incidence_list.openedges)
+                sub_tree, scs, tcs = tree_greedy(sub_list, log2_edge_sizes; nrepeat=nrepeat, α=α, temperature=temperature)
+                vi = contract_tree!(incidence_list, sub_tree, log2_edge_sizes, scs, tcs)
+                contraction_tree_nodes[tensors_list[vi]] = st2ct(sub_tree, tensors_list, contraction_tree_nodes)
+                flag = vi
+            end
+        end
+    end
+
+    return contraction_tree_nodes[tensors_list[flag]]
+end
+
+function st2ct(sub_tree::Union{ContractionTree, VT}, tensors_list::Dict{VT, Int}, contraction_tree_nodes::Vector{Union{ContractionTree, VT}}) where{VT}
+    if sub_tree isa ContractionTree
+        return ContractionTree(st2ct(sub_tree.left, tensors_list, contraction_tree_nodes), st2ct(sub_tree.right, tensors_list, contraction_tree_nodes))
+    else
+        return contraction_tree_nodes[tensors_list[sub_tree]]        
+    end
+end
+
+"""
+    optimize_exact_treewidth(optimizer, eincode, size_dict)
+
+Optimizing the contraction order via solve the exact tree width of the line graph corresponding to the eincode and return a `NestedEinsum` object.
+Check the docstring of `exact_treewidth_method` for detailed explaination of other input arguments.
+"""
+function optimize_exact_treewidth(optimizer::ExactTreewidth{GM}, code::EinCode{L}, size_dict::Dict) where {L,GM}
+    optimize_exact_treewidth(optimizer, getixsv(code), getiyv(code), size_dict)
+end
+function optimize_exact_treewidth(optimizer::ExactTreewidth{GM}, ixs::AbstractVector{<:AbstractVector}, iy::AbstractVector, size_dict::Dict{L,TI}) where {L, TI, GM}
+    if length(ixs) <= 2
+        return NestedEinsum(NestedEinsum{L}.(1:length(ixs)), EinCode(ixs, iy))
+    end
+    log2_edge_sizes = Dict{L,Float64}()
+    for (k, v) in size_dict
+        log2_edge_sizes[k] = log2(v)
+    end
+    # complete all open edges as a clique, connected with a dummy tensor
+    incidence_list = IncidenceList(Dict([i=>ixs[i] for i=1:length(ixs)] ∪ [(length(ixs) + 1 => iy)]))
+
+    α = optimizer.greedy_config.α
+    temperature = optimizer.greedy_config.temperature
+    nrepeat = optimizer.greedy_config.nrepeat
+    tree = exact_treewidth_method(incidence_list, log2_edge_sizes; α = α, temperature = temperature, nrepeat=nrepeat)
+
+    # remove the dummy tensor added for open edges
+    if isempty(iy)
```

Comment:
Avoid using if-else if not absolutely needed. It may reduce the test coverage.

Reply:
The if-else is replaced by `return isempty(iy) ? optcode : pivot_tree(optcode, length(ixs) + 1)`

Reply:
It does not make any difference. Why empty iy needs to be treated as a corner case, I do not understand.